app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,151 @@
+          hotspotManager.turnOnHotspot();
+          updateNotification(getString(R.string.hotspot_running), true);
+        }
+        break;
+
+      case ACTION_TURN_OFF_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          stopHotspot();
+        }
+        break;
+      default:
+        break;
+    }
+    return START_NOT_STICKY;
+  }
+
+  @Nullable @Override public IBinder onBind(Intent intent) {
+    return null;
+  }
+
+  private Notification buildForegroundNotification(String status, boolean showStopButton) {
+    Log.v(TAG, "Building notification " + status);
+    builder = new NotificationCompat.Builder(this);
+    builder.setContentTitle("Kiwix Hotspot").setContentText(status);
+    Intent targetIntent = new Intent(this, MainActivity.class);
+    targetIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
+    PendingIntent contentIntent =
+        PendingIntent.getActivity(this, 0, targetIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+    builder.setContentIntent(contentIntent)
+        .setSmallIcon(R.mipmap.kiwix_icon)
+        .setWhen(System.currentTimeMillis());
+
+    hotspotNotificationChannel();
+
+    if (showStopButton) {
+      Intent stopIntent = new Intent(ACTION_STOP);
+      PendingIntent stopHotspot =
+          PendingIntent.getBroadcast(this, 0, stopIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+      builder.addAction(R.drawable.ic_close_white_24dp, getString(R.string.tts_stop), stopHotspot);
+    }
+    return (builder.build());
+  }
+
+  private void updateNotification(String status, boolean stopAction) {
+    notificationManager.notify(HOTSPOT_NOTIFICATION_ID,
+        buildForegroundNotification(status, stopAction));
+  }
+
+  @RequiresApi(Build.VERSION_CODES.O)
+  void stopHotspot() {
+    hotspotManager.turnOffHotspot();
+    stopForeground(true);
+    stopSelf();
+    stopAndroidWebServer();
+    serverStateListener.serverStopped();
+  }
+
+  @Override
+  public void onDestroy() {
+    if (stopReceiver != null) {
+      unregisterReceiver(stopReceiver);
+    }
+    super.onDestroy();
+  }
+
+  private void hotspotNotificationChannel() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      NotificationChannel hotspotServiceChannel = new NotificationChannel(
+          Constants.HOTSPOT_SERVICE_CHANNEL_ID, getString(R.string.hotspot_service_channel_name),
+          NotificationManager.IMPORTANCE_DEFAULT);
+      hotspotServiceChannel.setDescription("Sample hotspot description");
+      hotspotServiceChannel.setSound(null, null);
+      builder.setChannelId(Constants.HOTSPOT_SERVICE_CHANNEL_ID);
+      notificationManager.createNotificationChannel(hotspotServiceChannel);
+    }
+  }
+
+  @RequiresApi(api = Build.VERSION_CODES.O)
+  public static boolean checkHotspotState(Context context) {
+    if (hotspotManager == null) {
+      Log.v("TAG", "hotspotManager initialized");
+      hotspotManager = new WifiHotspotManager(context);
```

Comment:
This is incredibly dangerous and undoubtedly a context leak, you create a static field with an activity context, this means the activity can never be garbage collected as this holds a reference to it.
WifiHotspotManager should be injected by dagger and be either an application scoped singelton or a service scoped singleton

Reply:
Fixed 7bbc07d3bff5295e128f046b386ec636bcc0834b